Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -0,0 +1,303 @@
+using Content.Shared.RCD.Components;
+using Content.Shared.Tag;
+using Robust.Shared.Audio;
+using Robust.Shared.Map;
+using Robust.Shared.Map.Components;
+using Robust.Shared.Network;
+using Robust.Shared.Serialization;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.RCD.Systems;
+
+public sealed class RCDSystem : EntitySystem
+{
+    [Dependency] private readonly ISharedAdminLogManager _adminLogger = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedChargesSystem _charges = default!;
+    [Dependency] private readonly SharedDoAfterSystem _doAfter = default!;
+    [Dependency] private readonly SharedInteractionSystem _interaction = default!;
+    [Dependency] private readonly IMapManager _mapMan = default!;
+    [Dependency] private readonly INetManager _net = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly TagSystem _tag = default!;
+    [Dependency] private readonly ITileDefinitionManager _tileDefMan = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+
+    private readonly int RCDModeCount = Enum.GetValues(typeof(RcdMode)).Length;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<RCDComponent, ExaminedEvent>(OnExamine);
+        SubscribeLocalEvent<RCDComponent, UseInHandEvent>(OnUseInHand);
+        SubscribeLocalEvent<RCDComponent, AfterInteractEvent>(OnAfterInteract);
+        SubscribeLocalEvent<RCDComponent, RCDDoAfterEvent>(OnDoAfter);
+    }
+
+    private void OnExamine(EntityUid uid, RCDComponent comp, ExaminedEvent args)
+    {
+        if (!args.IsInDetailsRange)
+            return;
+
+        var msg = Loc.GetString("rcd-component-examine-detail", ("mode", comp.Mode));
+        args.PushMarkup(msg);
+    }
+
+    private void OnUseInHand(EntityUid uid, RCDComponent comp, UseInHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        NextMode(uid, comp, args.User);
+        args.Handled = true;
+    }
+
+    private void OnAfterInteract(EntityUid uid, RCDComponent comp, AfterInteractEvent args)
+    {
+        if (args.Handled || !args.CanReach)
+            return;
+
+        var user = args.User;
+
+        var location = args.ClickLocation;
+        // Initial validity check
+        if (!location.IsValid(EntityManager))
+            return;
+
+        var gridIdOpt = location.GetGridUid(EntityManager);
+        if (!(gridIdOpt is EntityUid gridId) || !gridId.IsValid())
+        {
+            location = location.AlignWithClosestGridTile();
+            gridIdOpt = location.GetGridUid(EntityManager);
+            // Check if fixing it failed / get final grid ID
+            if (!(gridIdOpt is EntityUid gridId2) || !gridId2.IsValid())
+                return;
+            gridId = gridId2;
+        }
+
+        var mapGrid = _mapMan.GetGrid(gridId);
```

Comment:
You should TryComp for MapGrid rather than relying on the grid not being deleted (IsValid only checks the uid is valid and nothing about the state of the entity).

Reply:
the whole lot is sus im gonna change most of it